### PR TITLE
status quo: Barbara writes a runtime-agnostic library

### PR DESCRIPTION
**Pull request by zeenix**
This is a story from the async vision story session of April 16, 2021.

### CODE REVIEW

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+and enlists the help of Alan in doing so. Alan is already aware that [not all libraries in Rust work
+with all runtimes][nalirwwar]. Alan and Barbara start by writing a parser which works on
+`std::io::Read` and get their tests working with `String`s. After this they contemplate the question
+of how to accept a TCP connection.
+
+### Incompatible `AsyncRead` traits
+
+Alan asks Barbara what is the async equivalent is of `std::io::Read`, and Barbara sighs and says
+that there isn't one. Barbara brings up tokio's and the [futures crate]'s versions of `AsyncRead`.
+Barbara decides not to talk about `AsyncBufRead` for now.
```
Comment (rylev) [inline src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:32]: Just as a note: the upcoming story on `AsyncBuf{Read,Write}` will be good to link here.

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+
+Barbara and Alan decide to use the future's `AsyncRead` for no other reason other than it is
+runtime-agnostic. Barbara tells Alan not to worry as they can translate between the two. With some
+effort (**TODO:** maybe link a story about `Pin`? which one?) they convert their parser to using
```
Comment (rylev) [inline src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:36]: I think we don't have a good story for `AsyncRead` yet. We do mention `Pin` difficulties in a few places like in [here](https://rust-lang.github.io/wg-async-foundations/vision/status_quo/barbara_needs_async_helpers.html) and [here](https://github.com/rust-lang/wg-async-foundations/blob/master/src/vision/status_quo/alan_hates_writing_a_stream.md), but as we discussed today in the session on `AsyncRead`/`AsyncWrite`, we'll probably want a dedicated story for this.

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+### Task spawning
+
+After getting the TCP handling part working, they now want to spawn tasks for handling each incoming
+TCP connection. Again, to their disappointment, they find that there's no runtime-agnostic way to do
+that.
+
+Unsure on how to do this, they do some searching and find the [`agnostik`] crate. They reject it
+because this only supports N number of runtimes and their custom runtime is not one of them.
+However it gives them the idea to provide a trait for specifying how to spawn tasks on the runtime.
+Barbara points out that this has disadvantage of working against orphan rules meaning that either
```
Comment (rylev) [inline src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:55]: We might want to either expand on this here or perhaps better provide a footnote explaining this. There's also now an issue tracking this: #180

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+they have to implement the trait for all known runtimes (defeating the purpose of the exercise) or
+force the user to use new types.
+
+They punt on this question by implementing the trait for each of the known runtimes. They're
+disappointed that this means their library actually isn't runtime agnostic.
+
+### The need for timers
```
Comment (rylev) [inline src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:62]: I'd like to mention that they could add a timer method to their runtime trait and much like spawning require every runtime to provide an implementation, but they decide not to do this to try to keep this trait as simple as possible.

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+
+### **What are the morals of the story?**
+
+* People have to roll their own implementations which can lead to often subtle differences between
+  runtimes (For example TCPListeners in `async-std` and `tokio`).
+* Orphan rules and no standard traits guarantee that a truly agnostic library is not possible.
+* Takes way more time than writing synchronous protocols.
+* It's a hard goal to achieve.
+* Leads to poorer APIs sometimes (both in ease of use and **performance**).
+* More thought is needed to write such libraries.
```
Comment (rylev) [inline src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:96]:
```suggestion
* More API design considerations need to go into making an generic async library than a generic sync library.
```

**Review by rylev — COMMENTED** [review]
Good start! 🧡

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md (hunk 1 above)
Comment (zeenix) [reply src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:32]: Sure, when there is a link available. :)

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md (hunk 3 above)
Comment (zeenix) [reply src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:55]: I think a link to this issue would suffice here but feel free to suggest what further info should be provided here.

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md (hunk 4 above)
Comment (zeenix) [reply src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md:62]: That's already there, isn't it? Or do you mean I should change the wording?

File: src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md (hunk 2 above)
Comment (zeenix) [reply src/vision/status_quo/barbara_writes_a_runtime_agnostic_lib.md]: This one was about Pin difficulties specifically so I've added links to those stories now (thanks).

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**